src/lookup.h

```diff
@@ -0,0 +1,267 @@
+/*
+ * Copyright The async-profiler authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef _LOOKUP_H
+#define _LOOKUP_H
+
+#include "demangle.h"
+#include "dictionary.h"
+#include "index.h"
+#include "profiler.h"
+#include "vmEntry.h"
+#include "vmStructs.h"
+#include <jvmti.h>
+
+class MethodInfo {
+  public:
+    MethodInfo() : _mark(false), _key(0) {
+    }
+
+    bool _mark;
+    u32 _key;
+    u32 _class;
+    u32 _file;
+    u32 _name;
+    u32 _system_name;
+    u32 _sig;
+    jint _modifiers;
+    jint _line_number_table_size;
+    jvmtiLineNumberEntry* _line_number_table;
+    FrameTypeId _type;
+
+    jint getLineNumber(jint bci) {
+        if (_line_number_table_size == 0) {
+            return 0;
+        }
+
+        int i = 1;
+        while (i < _line_number_table_size && bci >= _line_number_table[i].start_location) {
+            i++;
+        }
+        return _line_number_table[i - 1].line_number;
+    }
+};
+
+class MethodMap : public std::map<jmethodID, MethodInfo> {
+  public:
+    MethodMap() {
+    }
+
+    ~MethodMap() {
+        jvmtiEnv* jvmti = VM::jvmti();
+        for (const_iterator it = begin(); it != end(); ++it) {
+            jvmtiLineNumberEntry* line_number_table = it->second._line_number_table;
+            if (line_number_table != NULL) {
+                jvmti->Deallocate((unsigned char*)line_number_table);
+            }
+        }
+    }
+
+    size_t usedMemory() {
+        size_t bytes = 0;
+        for (const_iterator it = begin(); it != end(); ++it) {
+            bytes += sizeof(jmethodID) + sizeof(MethodInfo);
+            bytes += it->second._line_number_table_size * sizeof(jvmtiLineNumberEntry);
+        }
+        return bytes;
+    }
+};
+
+class Lookup {
+  public:
+    MethodMap* _method_map;
+    // Dictionary is thread-safe and can be safely shared, as opposed to Index
+    Dictionary* _classes;
+    Index* _packages;
+    Index* _symbols;
+
+  private:
+    JNIEnv* _jni;
+    bool _owns_packages = false;
+    bool _owns_symbols = false;
+
+    void fillNativeMethodInfo(MethodInfo* mi, const char* name, const char* lib_name) {
+        if (lib_name == NULL) {
+            mi->_file = _classes->lookup("");
+        } else if (lib_name[0] == '[' && lib_name[1] != 0) {
+            mi->_file = _classes->lookup(lib_name + 1, strlen(lib_name) - 2);
+        } else {
+            mi->_file = _classes->lookup(lib_name);
+        }
+
+        mi->_modifiers = 0x100;
+        mi->_line_number_table_size = 0;
+        mi->_line_number_table = NULL;
+
+        mi->_system_name = _symbols->indexOf(name);
```

Comment:
This pollutes symbol table in JFR, even though `_system_name` is never used there.

Reply:
520d2f1cf10720c5561880d479f990ba512b6baf